src/sync/scheduler.ts

```diff
@@ -0,0 +1,146 @@
+import { Repository, Subscription } from "models/subscription";
+import { Task, TaskType } from "~/src/sync/sync.types";
+import { RepoSyncState } from "models/reposyncstate";
+import { getTargetTasks } from "~/src/sync/installation";
+import { createInstallationClient } from "utils/get-github-client-config";
+import Logger from "bunyan";
+import { booleanFlag, BooleanFlags } from "config/feature-flags";
+
+const getCursorKey = (type: TaskType) => `${type}Cursor`;
+const getStatusKey = (type: TaskType) => `${type}Status`;
+
+// These numbers were obtained experimentally by syncing a large customer with 60K quota from GitHub
+const RATE_LIMIT_QUOTA_PER_TASK_RESERVE = 500;
+const MAX_NUMBER_OF_SUBTASKS = 100;
+const SUBTASKS_POLL_MAX_SIZE = MAX_NUMBER_OF_SUBTASKS * 10;
+
+/**
+ *
+ * @param subscription
+ * @param logger
+ * @param mainTask
+ * @param otherTasks - the array will be shuffled after the call, BEWARE!
+ */
+const calculateTasksUsingGitHubRateLimitQuota = async (subscription: Subscription, logger: Logger, mainTask: Task, otherTasks: Task[]) => {
+	try {
+		const metrics = {
+			trigger: "ratelimit_check_backfill"
+		};
+		const rateLimitResponse = await (await createInstallationClient(subscription.gitHubInstallationId, subscription.jiraHost, metrics, logger, subscription.gitHubAppId)).getRateLimit();
+		const rateLimitData = rateLimitResponse.data;
+
+		const availQuota = Math.min(rateLimitData.resources.core.remaining, rateLimitData.resources.graphql.remaining);
+		// We need to reserve some quota for the main task, too
+		const availQuotaForSubtasks = Math.max(0, availQuota - RATE_LIMIT_QUOTA_PER_TASK_RESERVE);
+		const allowedSubtasks = Math.floor(availQuotaForSubtasks / RATE_LIMIT_QUOTA_PER_TASK_RESERVE);
+
+		const nSubTasks = Math.min(allowedSubtasks, MAX_NUMBER_OF_SUBTASKS);
+
+		logger.info({ nSubTasks, rateLimitData }, "Using subtasks: " + nSubTasks);
+
+		if (nSubTasks === 0) {
+			return [mainTask];
+		}
+
+		return [mainTask, ...(
+			otherTasks.slice(0, nSubTasks).sort(() => Math.random() - 0.5)
+		)];
+	} catch (err) {
+		logger.warn({ err }, "Cannot determine rate limit, return only main task");
+		return [mainTask];
+	}
+};
+
+const mapSyncStateToTasks = (tasks: TaskType[], syncState: RepoSyncState): Task[] => {
+	const ret: Task[] = [];
```

Comment:
What is ret? A little hard to determine what this is from the name. Maybe `syncStateTask` or something along those lines?